salt/proxy/onyx.py

```diff
@@ -0,0 +1,609 @@
+    return ret
+
+
+def disable():
+    '''
+        Shortcut to run `disable` on switch
+
+        .. code-block:: bash
+
+            salt '*' onyx.cmd disable
+        '''
+    try:
+        ret = sendline('disable')
+    except TerminalException as e:
+        log.error(e)
+        return 'Failed to "configure terminal"'
+    return ret
+
+
+def grains():
+    '''
+    Get grains for proxy minion
+
+    .. code-block: bash
+
+        salt '*' onyx.cmd grains
+    '''
+    if not DETAILS['grains_cache']:
+        ret = system_info()
+        log.debug(ret)
+        DETAILS['grains_cache'].update(ret)
+    return {'onyx': DETAILS['grains_cache']}
+
+
+def grains_refresh():
+    '''
+    Refresh the grains from the proxy device.
+
+    .. code-block: bash
+
+        salt '*' onyx.cmd grains_refresh
+    '''
+    DETAILS['grains_cache'] = {}
+    return grains()
+
+
+def get_user(username):
+    '''
+    Get username line from switch
+
+    .. code-block: bash
+
+        salt '*' onyx.cmd get_user username=admin
+    '''
+    try:
+        enable()
+        configure_terminal()
+        cmd_out = sendline('show running-config | include "username {0} password 7"'.format(username))
+        cmd_out.split('\n')
+        user = cmd_out[1:-1]
+        configure_terminal_exit()
+        disable()
+        return user
+    except TerminalException as e:
+        log.error(e)
+        return 'Failed to get user'
+
+
+def get_roles(username):
+    '''
+    Get roles that the username is assigned from switch
+
+    .. code-block: bash
+
+        salt '*' onyx.cmd get_roles username=admin
+    '''
+    info = sendline('show user-account {0}'.format(username))
+    roles = re.search(r'^\s*roles:(.*)$', info, re.MULTILINE)
+    if roles:
+        roles = roles.group(1).strip().split(' ')
+    else:
+        roles = []
+    return roles
+
+
+def check_password(username, password, encrypted=False):
+    '''
+    Check if passed password is the one assigned to user
+
+    .. code-block: bash
+
+        salt '*' onyx.cmd check_password username=admin password=admin
+        salt '*' onyx.cmd check_password username=admin \\
+            password='$5$2fWwO2vK$s7.Hr3YltMNHuhywQQ3nfOd.gAPHgs3SOBYYdGT3E.A' \\
+            encrypted=True
+    '''
+    hash_algorithms = {'1': 'md5',
+                       '2a': 'blowfish',
+                       '5': 'sha256',
+                       '6': 'sha512', }
+    user = get_user(username)
+    password_line = user.split('\n')
+    password_line = password_line[1]
+    if not password_line:
+        return None
+    if '!!' in password_line:
+        return False
+    cur_hash = re.search(r'(\$[0-6](?:\$[^$ ]+)+)', str(password_line)).group(0)
+    if encrypted is False:
+        hash_type, cur_salt, hashed_pass = re.search(r'^\$([0-6])\$([^$]+)\$(.*)$', str(cur_hash)).groups()
+        new_hash = gen_hash(crypt_salt=cur_salt, password=password, algorithm=hash_algorithms[hash_type])
```

Comment:
Be advised that this will raise an exception on Windows.

Reply:
I have removed this function for now!